app/models/amazon_agent_manager.rb

```diff
@@ -0,0 +1,393 @@
+    bucket = s3.bucket(ssa_bucket)
+    return false unless bucket.exists?
+
+    obj_id = heartbeat_prefix + agent_id
+    obj = bucket.object(obj_id)
+    return false unless obj.exists?
+
+    last_beat_stamp = obj.last_modified
+    _log.debug("#{obj.key}: Last heartbeat time stamp: #{last_beat_stamp}")
+
+    Time.now.utc - last_beat_stamp < interval && ec2.instance(agent_id).state.name == "running" ? true : false
+  rescue => err
+    _log.error("#{agent_id}: #{err.message}")
+    false
+  end
+
+  def deploy_agent
+    _log.info("Deploying agent ...")
+    @deploying = true
+
+    kp = get_key_pair
+    security_group_id = create_security_group
+    create_setting_yaml
+    data = create_userdata
+    zone_name = ec2.client.describe_availability_zones.availability_zones[0].zone_name
+    subnets = get_subnets(zone_name)
+    raise "No subnet_id is available for #{zone_name}!" if subnets.length.zero?
+    create_profile
+
+    instance = ec2.create_instances({
+      image_id: get_agent_image_id,
+      min_count: 1,
+      max_count: 1,
+      key_name: kp.name,
+      security_group_ids: [security_group_id],
+      user_data: data,
+      instance_type: 't2.micro',
+      placement: {
+        availability_zone: zone_name
+      },
+      subnet_id: subnets[0].subnet_id,
+      iam_instance_profile: {
+        name: profile_name
+      },
+      tag_specifications: [{
+        resource_type: "instance",
+        tags: [{
+          key: "Name",
+          value: MIQ_SSA
+        }]
+      }]
+    })
+    ec2.client.wait_until(:instance_status_ok, {instance_ids: [instance[0].id]})
+
+    instance[0].id
+  end
+
+  private
+
+  def role_exists?(role_name)
+    role = iam.role(role_name)
+    role.role_id
+    true
+  rescue ::Aws::IAM::Errors::NoSuchEntity
+    false
+  end
+
+  def find_or_create_role(role_name = MIQ_SSA)
+    return iam.role(role_name) if role_exists?(role_name)
+
+    # Policy Generator:
+    policy_doc = {
+      :Version => "2012-10-17",
+      :Statement => [{
+        :Effect => "Allow",
+        :Principal => {:Service => "ec2.amazonaws.com"},
+        :Action => "sts:AssumeRole"
+      }]
+    }
+
+    role = iam.create_role(
+      role_name: role_name,
+      assume_role_policy_document: policy_doc.to_json
+    )
+
+    # grant all priviledges
+    role.attach_policy({
+      policy_arn: 'arn:aws:iam::aws:policy/AmazonS3FullAccess'
+    })
+
+    role.attach_policy({
+      policy_arn: 'arn:aws:iam::aws:policy/AmazonEC2FullAccess'
+    })
+
+    role.attach_policy({
+      policy_arn: 'arn:aws:iam::aws:policy/AmazonSQSFullAccess'
+    })
+
+    role
+  end
+
+  def create_profile(profile_name = MIQ_SSA, role_name = MIQ_SSA)
+    ssa_profile = iam.instance_profile(profile_name)
+    ssa_profile = iam.create_instance_profile(instance_profile_name: profile_name) unless ssa_profile.exists?
+
+    find_or_create_role(role_name)
+    ssa_profile.add_role(role_name: role_name) if ssa_profile.roles.zero?
```

Comment:
Is `roles` here a number or an Array?  I would have expected an Array, but the .zero? implies it's not, and it's weird to have a noun return a number without a qualifying word like "_count"

Reply:
It's an Array. Rubocop recommends to use zero? rather than .count == 0

Reply:
Array#zero? doesn't exist, so this is broken.  Testing should have caught this because you can't get past this point otherwise.

Even so, using `.zero?` is a smell, because for Array's it's preferable to use `.empty?`